### PR TITLE
Fixes nav pan and zoom bug

### LINKED ISSUE #278: Nav bar panning zoom-in bug
before

To see, pan left repeatedly when brush is at LHS of domain.

after

Then zoom in.

### PR DESCRIPTION
**Pull request by djmiley**
Resolves #278

### CODE REVIEW

**Comment by djmiley** [comment]
Need to change this!!!

**Comment by djmiley** [comment]
Need to add some unit tests

File: src/assets/js/util/domain/strictlyWithinDateExtentOfData.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
+(function(d3, fc, sc) {
+    'use strict';
+
+    function domainAfterStartOfData(domain, data) {
+        var earliestViewedTime = domain[0].getTime();
+        var earliestDatumTime = data[0].date.getTime();
```
Comment (tlsim) [inline src/assets/js/util/domain/strictlyWithinDateExtentOfData.js]: This assumes the data is sorted by date. Can you use something from https://github.com/mbostock/d3/wiki/Arrays ?

File: src/assets/js/util/domain/strictlyWithinDateExtentOfData.js [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    }
+
+    function domainBeforeEndOfData(domain, data) {
+        var latestViewedTime = domain[1].getTime();
+        var lastDatumTime = data[data.length - 1].date.getTime();
+        return latestViewedTime < lastDatumTime;
+    }
+
+    sc.util.domain.strictlyWithinDateExtentOfData = function(domain, data) {
+        return domainAfterStartOfData(domain, data) && domainBeforeEndOfData(domain, data);
```
Comment (tlsim) [inline src/assets/js/util/domain/strictlyWithinDateExtentOfData.js]: Might be easier to first get the date extent of data, then compare the extents.

File: src/assets/js/behavior/zoom.js [hunk 3]
```diff
@@ -58,6 +58,8 @@
                     }
                     if ((panned && allowPan) || (zoomed && allowZoom)) {
                         dispatch.zoom(domain);
+                    } else {
+                        dispatch.zoom();
```
Comment (djmiley) [inline src/assets/js/behavior/zoom.js]: Have a think about this approach.

File: src/assets/js/chart/nav.js [hunk 4]
```diff
@@ -65,7 +65,11 @@
                 .trackingLatest(selection.datum().trackingLatest)
                 .allowPan(false)
                 .on('zoom', function(domain) {
-                    dispatch.viewChange(domain);
+                    if (domain) {
+                        dispatch.viewChange(domain);
+                    } else if (!sc.util.domain.strictlyWithinDateExtentOfData(model.viewDomain, model.data)) {
+                        dispatch.viewChange(model.viewDomain);
+                    }
```
Comment (djmiley) [inline src/assets/js/chart/nav.js]: Likewise.

File: src/assets/js/chart/nav.js [hunk 5]
```diff
@@ -65,7 +65,11 @@
                 .trackingLatest(selection.datum().trackingLatest)
                 .allowPan(false)
                 .on('zoom', function(domain) {
-                    dispatch.viewChange(domain);
+                    if (domain) {
```
Comment (tlsim) [inline src/assets/js/chart/nav.js]: Feels odd that on 'zoom' a domain might not be returned. Let's take another look tomorrow.

Comment (djmiley) [inline src/assets/js/chart/nav.js]: :+1:

**Comment by djmiley** [comment]
Caused by this change

https://github.com/ScottLogic/d3fc-showcase/commit/580e45cd585d5cbae8b66180e1278080d7206410

**Pull request merged**